recipe/2023-11-validate-template-path.md

```diff
@@ -0,0 +1,96 @@
+
+
+#### Validate module sources as part of the CreateOrUpdateEnvironment API endpoint 
+Right now, we only return an error at deployment time. We'd like to return an error while the recipe is being registered to the environment. To do so, we'll add validation as part of the `CreateOrUpdateEnvironment` API. This requires no new API changes, just an update to how we handle environment conversion.
+
+We already have a check for local path module sources [here](https://github.com/radius-project/radius/blob/40c91fdc3a4dd3ac04906094dc8302f7232d700d/pkg/corerp/api/v20231001preview/environment_conversion.go#L303). Instead of just rejecting local module source paths, we'll only accept Terraform registries and HTTP URLs and reject everything else.  
+
+#### Validation steps
+Terraform registries: 
+Terraform has code to check whether a module source is a Terraform recipe as part of the `terraform-registry-address` [package](https://github.com/hashicorp/terraform-registry-address/blob/main/module.go#L44). We will use this package to validate Terraform module registries. 
+
+HTTP URLs: 
+Terraform treats HTTP URLs as remote addresses and doesn't have functionality for checking HTTP URLs in isolation. We will need to implement our own validation for HTTP URLs. Go provides validation for URIs as part of the `url` [package](https://pkg.go.dev/net/url#ParseRequestURI), but we will likely need additional checks to ensure that a `.zip`/`.tar`/etc suffix is present as well. 
+
+Any module source that is not a Terraform registry or an HTTP URL will be treated as unsupported. We will return a `NewClientErrInvalidRequest` [error](https://github.com/radius-project/radius/blob/main/pkg/armrpc/api/v1/error.go#L64) in this case. The error message will indicate that we only support Terraform registries and HTTP URLs currently. 
+
+## Alternatives considered
+
+1. Adding a `ValidateTemplateKind` API endpoint (similar to `GetRecipeMetadata`).
+This was rejected because the logic could be added to the existing `CreateOrUpdateEnvironment` API endpoint instead. That way, there's less code to add and we can return a validation error before we save any environment information. 
+2. Support all Terraform module sources instead of restricting supported sources.
+Support for other modules will be done on an as-needed customer basis. This needs to be investigated more to see which module sources need more support and if it's viable to support all of them easily. 
+
+
+## Test plan
+
+Unit testing:
+- Add tests for accepted module sources: HTTP URLs and Terraform module registries 
+  - For HTTP URLs, make sure to test negative cases as the Go URL package isn't very robust 
+- Add tests for module sources that are not accepted (validate error thrown): local paths, Github, S3, etc 
+- Add tests in environment conversion for different module sources (and validate errors/successes)
+
+## Security
+
```

Comment:
The main thing I'd call out for security is that **any** parsing code that operates on untrusted input can cause security issues, usually by introducing an infinite loop. 

We should avoid writing our own parsers, and rely on libraries. This is why it's important that we're using the built-in Go URL type.